presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java

```diff
@@ -44,21 +45,9 @@ public static Block union(
         int leftArrayCount = leftArray.getPositionCount();
         int rightArrayCount = rightArray.getPositionCount();
         TypedSet typedSet = new TypedSet(type, leftArrayCount + rightArrayCount, "array_union");
-        BlockBuilder distinctElementBlockBuilder = type.createBlockBuilder(null, leftArrayCount + rightArrayCount);
-        appendTypedArray(leftArray, type, typedSet, distinctElementBlockBuilder);
-        appendTypedArray(rightArray, type, typedSet, distinctElementBlockBuilder);
-
-        return distinctElementBlockBuilder.build();
-    }
-
-    private static void appendTypedArray(Block array, Type type, TypedSet typedSet, BlockBuilder blockBuilder)
-    {
-        for (int i = 0; i < array.getPositionCount(); i++) {
-            if (!typedSet.contains(array, i)) {
-                typedSet.add(array, i);
-                type.appendTo(array, i, blockBuilder);
-            }
-        }
+        appendTypedArray(leftArray, typedSet);
+        appendTypedArray(rightArray, typedSet);
+        return typedSet.getBlock();
```

Comment:
Note `ArrayUnion` should a cached `PageBuilder` to avoid allocate a new `BlockBuilder` in each invocation. 

This optimization cannot work with the cached `PageBuilder` solution. 

More discussions can be found in https://github.com/prestodb/presto/issues/12033

Reply:
As discussed in #11984 array_union needs the NoCopyTypedSet or JsonUtil#HashTable to support adding more than one blocks if we go through the first optimization. And we need to evaluate the performance using that solution. This PR shows the second optimization does show some gain for the array_union case. I'm waiting for people to chime in to decide which way to go, but I think we can keep both NoCopyTypedSet/ JsonUtil#HashTable and this modified TypedSet and the callers can decide which class to use.